### ISSUE TITLE
Parsing error when property sheet files use CRLF line endings

### ISSUE THREAD

**Issue by brupelo** [tree-sitter/tree-sitter#357]
Hi!

First time trying to use tree-sitter... my goal was building & using these existing grammars:

https://github.com/tree-sitter/tree-sitter-python
https://github.com/tree-sitter/tree-sitter-javascript

with the python bindings https://github.com/tree-sitter/py-tree-sitter.

Unfortunately, everything is failing here :) . 

I've tried generating the source of those grammars with the tree-sitter exes (x86 & x64) provided in the releases sections of the repo but those exes will give me 'nvalid identifier ' messages when running `tree-sitter generate` in the root of those grammar repos.

Then I've tried using a built manually tree-sitter executable using cargo and it gives the same error, what does that cryptic error means btw?

After giving up of building those grammar myself I've decided to use directly the bundled grammar source provided in the repos directly with the py-tree-sitter example: 

    from tree_sitter import Language, Parser

    def build():
        Language.build_library(
            # Store the library in the `build` directory
            'build/my-languages.pyd',

            # Include one or more languages
            [
                # 'vendor/tree-sitter-go',
                # 'vendor/tree-sitter-javascript',
                'tree-sitter-python'
            ]
        )

    def main():
        PY_LANGUAGE = Language('build/my-languages.pyd', 'python')
        parser = Parser()
        parser.set_language(PY_LANGUAGE)

        # import textwrap
        # tree = parser.parse(textwrap.dedent("""\
        #     def foo():
        #         if bar:
        #             baz()
        # """).encode("utf-8"))

        tree = parser.parse(bytes("""
        def foo():
            if bar:
                baz()
        """, "utf8"))

    if __name__ == '__main__':
        build()
        main()

but I'd get the below error:

    (py364_32) D:\sources\personal\python\pyblime\pyblime\modes\py3antlr4book\0_test\mcve>python t.py
    parser.c
       Creating library build/my-languages.lib and object build/my-languages.exp
    parser.obj : error LNK2001: unresolved external symbol _tree_sitter_python_external_scanner_create
    parser.obj : error LNK2001: unresolved external symbol _tree_sitter_python_external_scanner_serialize
    parser.obj : error LNK2001: unresolved external symbol _tree_sitter_python_external_scanner_deserialize
    parser.obj : error LNK2001: unresolved external symbol _tree_sitter_python_external_scanner_scan
    parser.obj : error LNK2001: unresolved external symbol _tree_sitter_python_external_scanner_destroy
    build/my-languages.pyd : fatal error LNK1120: 5 unresolved externals
    Traceback (most recent call last):
      File "d:\software\python364_32\Lib\distutils\_msvccompiler.py", line 519, in link
        self.spawn([self.linker] + ld_args)
      File "d:\software\python364_32\Lib\distutils\_msvccompiler.py", line 542, in spawn
        return super().spawn(cmd)
      File "d:\software\python364_32\Lib\distutils\ccompiler.py", line 909, in spawn
        spawn(cmd, dry_run=self.dry_run)
      File "d:\software\python364_32\Lib\distutils\spawn.py", line 38, in spawn
        _spawn_nt(cmd, search_path, dry_run=dry_run)
      File "d:\software\python364_32\Lib\distutils\spawn.py", line 81, in _spawn_nt
        "command %r failed with exit status %d" % (cmd, rc))
    distutils.errors.DistutilsExecError: command 'C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\VC\\BIN\\link.exe' failed with exit status 1120

    During handling of the above exception, another exception occurred:

    Traceback (most recent call last):
      File "t.py", line 38, in <module>
        build()
      File "t.py", line 13, in build
        'tree-sitter-python'
      File "D:\virtual_envs\py364_32\lib\site-packages\tree_sitter\__init__.py", line 65, in build_library
        compiler.link_shared_object(object_paths, output_path)
      File "d:\software\python364_32\Lib\distutils\ccompiler.py", line 717, in link_shared_object
        extra_preargs, extra_postargs, build_temp, target_lang)
      File "d:\software\python364_32\Lib\distutils\_msvccompiler.py", line 522, in link
        raise LinkError(msg)
    distutils.errors.LinkError: command 'C:\\Program Files (x86)\\Microsoft Visual Studio 14.0\\VC\\BIN\\link.exe' failed with exit status 1120

Last but not least... I've also tried installing tree-sitter using npm but I've got a bunch of errors... :)

      
    > tree-sitter@0.15.1 install D:\sources\personal\python\pyblime\pyblime\modes\py3antlr4book\0_test\mcve\node_modules\tree-sitter
    > prebuild-install || node-gyp rebuild

    prebuild-install WARN install No prebuilt binaries found (target=12.4.0 runtime=node arch=x64 libc= platform=win32)

    (py364_32) D:\sources\personal\python\pyblime\pyblime\modes\py3antlr4book\0_test\mcve\node_modules\tree-sitter>if not defined npm_config_node_gyp (node "C:\Users\KneDa\AppData\Roaming\nvm\v12.4.0\node_modules\npm\node_modules\npm-lifecycle\node-gyp-bin\\..\..\node_modules\node-gyp\bin\node-gyp.js" rebuild )  else (node "C:\Users\KneDa\AppData\Roaming\nvm\v12.4.0\node_modules\npm\node_modules\node-gyp\bin\node-gyp.js" rebuild ) 
    gyp ERR! configure error 
    gyp ERR! stack Error: Command failed: D:\virtual_envs\py364_32\Scripts\python.EXE -c import sys; print "%s.%s.%s" % sys.version_info[:3];
    gyp ERR! stack   File "<string>", line 1
    gyp ERR! stack     import sys; print "%s.%s.%s" % sys.version_info[:3];
    gyp ERR! stack                                ^
    gyp ERR! stack SyntaxError: invalid syntax
    gyp ERR! stack 
    gyp ERR! stack     at ChildProcess.exithandler (child_process.js:290:12)
    gyp ERR! stack     at ChildProcess.emit (events.js:200:13)
    gyp ERR! stack     at maybeClose (internal/child_process.js:1021:16)
    gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:283:5)
    gyp ERR! System Windows_NT 6.1.7601
    gyp ERR! command "C:\\Program Files\\nodejs\\node.exe" "C:\\Users\\KneDa\\AppData\\Roaming\\nvm\\v12.4.0\\node_modules\\npm\\node_modules\\node-gyp\\bin\\node-gyp.js" "rebuild"
    gyp ERR! cwd D:\sources\personal\python\pyblime\pyblime\modes\py3antlr4book\0_test\mcve\node_modules\tree-sitter
    gyp ERR! node -v v12.4.0
    gyp ERR! node-gyp -v v3.8.0
    gyp ERR! not ok 
    npm WARN enoent ENOENT: no such file or directory, open 'D:\sources\personal\python\pyblime\pyblime\modes\py3antlr4book\0_test\mcve\package.json'
    npm WARN mcve No description
    npm WARN mcve No repository field.
    npm WARN mcve No README data
    npm WARN mcve No license field.

    npm ERR! code ELIFECYCLE
    npm ERR! errno 1
    npm ERR! tree-sitter@0.15.1 install: `prebuild-install || node-gyp rebuild`
    npm ERR! Exit status 1
    npm ERR! 
    npm ERR! Failed at the tree-sitter@0.15.1 install script.
    npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

    npm ERR! A complete log of this run can be found in:
    npm ERR!     C:\Users\KneDa\AppData\Roaming\npm-cache\_logs\2019-06-09T23_27_23_973Z-debug.log

It feels like is using some python2 syntax over there...

Summing up, this is my frst time trying to use tree-sitter but every of my attempts have failed miserably... could you advice? I'm on windows7 and I've got python3.6.3x86 + {vs2015 & vs2017} + nvm/latest node + rust.

Thx in advance.

**Comment by brupelo** [comment]
Ok, let me try again to provide a more concise bug-report:

1) **Build latest tree-sitter**

- mkdir d:\mcve && cd d:\mcve
- git clone https://github.com/tree-sitter/tree-sitter
- cd tree-sitter && cargo build --release
- cd target\release && set path=%cd%;%path%
- tree-sitter --version => tree-sitter 0.15.3 (07e49c4c59ff6c9f71fb896c15b49d47c40f0a1f)

2) **Check tree-sitter-python grammar**

- cd d:\mcve
- git clone https://github.com/tree-sitter/tree-sitter-python && cd tree-sitter-python
- python -c "open('test.py', 'w').write('def foo():\n    if bar:\n        baz()')"
- tree-sitter parse test.py

    (py364_32) D:\mcve\tree-sitter-python>tree-sitter parse test.py
    (module [0, 0] - [2, 13]
      (function_definition [0, 0] - [2, 13]
        (identifier [0, 4] - [0, 7])
        (parameters [0, 7] - [0, 9])
        (if_statement [1, 4] - [2, 13]
          (identifier [1, 7] - [1, 10])
          (expression_statement [2, 8] - [2, 13]
            (call [2, 8] - [2, 13]
              (identifier [2, 8] - [2, 11])
              (argument_list [2, 11] - [2, 13]))))))

So far so good but now let's regenerate the grammar... 

    (py364_32) D:\mcve\tree-sitter-python>mv src src_backup

    (py364_32) D:\mcve\tree-sitter-python>tree-sitter generate
    'nvalid identifier '

    (py364_32) D:\mcve\tree-sitter-python>tree-sitter parse test.py
    Failed to load language in current directory
    Details:
      Parser compilation failed.
    Stdout: parser.c
       Creating library parser.lib and object parser.exp
    parser.obj : error LNK2001: unresolved external symbol _tree_sitter_python_external_scanner_create
    parser.obj : error LNK2001: unresolved external symbol _tree_sitter_python_external_scanner_destroy
    parser.obj : error LNK2001: unresolved external symbol _tree_sitter_python_external_scanner_scan
    parser.obj : error LNK2001: unresolved external symbol _tree_sitter_python_external_scanner_serialize
    parser.obj : error LNK2001: unresolved external symbol _tree_sitter_python_external_scanner_deserialize
    C:\Users\KneDa\.tree-sitter\bin\python.dll : fatal error LNK1120: 5 unresolved externals

    Stderr:

Mmm, I don't understand... what's the problem here? :O

**Comment by brupelo** [comment]
Hold on, this looks fishy:

    (py364_32) D:\mcve\tree-sitter-python>tree-sitter generate
    'nvalid identifier '
    
    (py364_32) D:\mcve\tree-sitter-python>tree /f c:\Users\KneDa\.tree-sitter
    Folder PATH listing
    Volume serial number is 5A46-525F
    C:\USERS\KNEDA\.TREE-SITTER
    └───bin

As you can see that bin folder is empty... can this be related to the fact I'm using a virtualenv? Shouldn't be that user folder populated with something?

**Comment by brupelo** [comment]
I've been debugging for a while and I think I've discovered the main root of all the problems i've reported in this issue:

The reason why the message was 'nvalid identifier ' was mainly because there were windows line-endings crlf = '\r\n' but on properties.rs you're using:

    return Err(Error(format!("Invalid identifier '{}'", name)));

so that Error won't print properly invisible characters such as '\r'.

I've changed my git config to:

    git config --global core.autocrlf false

cloned again the tree-sitter-python repo and now it seems the src is generated properly.

But this is definitely a bug:

1) You should deal properly with windows/unix/macosx line ending (ideal solution)
2) You should improve the error report so the user will know straightaway what could be
3) You should document about this fact somewhere

**Comment by maxbrunsfeld** [comment]
Thanks for the report, and nice job narrowing down the problem. I agree this is definitely a bug. I believe that it may be caused by a bug in one of Tree-sitter's dependencies: [rsass](https://github.com/kaj/rsass). This is a library we use to parse CSS files for the purpose of generating syntax-highlighting property sheets (a fairly new feature which I haven't documented yet).

When I get a chance, I'll try to look into this further and see if there's an easy fix. If not, we can work around it by converting CRLF -> LF before calling into rsass.

**Comment by maxbrunsfeld** [comment]
It seems like you've identified a separate issue with [`py-tree-sitter`](https://github.com/tree-sitter/py-tree-sitter). Could you open a separate issue on that repository describing the linker error?

**Comment by brupelo** [comment]
@maxbrunsfeld Hey Max! nice to meet you... Sure, when I'm back home I'll continue testing/toying around and I'll open a separate issue on that repo with proper concise details. 

Btw, just to let you know I'm quite excited about using this project of yours... not sure if you've read 
 had the chance to read my other thread about "tree-sitter support" but I must to say after having tried few dozens of parsing libraries I think this one may be the one I'll feel comfortable with. 

For instance, one of my last attempts was using antlr4 with a glsl grammar and parsing a "simple" 28kb file of shader code would take 1.9s!!! That's just crazy, it's couple of order of magnitude that even compiling the shader :/ ... Of course, I was using a python-runtime instead of a cpp-runtime compiled as a python extension but even so... Not sure antlr would be suitable to achieve all my goals (although the ammount of existing grammars for antlr is really nice, check [here](https://github.com/antlr/grammars-v4)).

Hopefully this week I'll be able to integrate tree-sitter on a QScintilla widget and adding basic features such as syntax highlighting & code folding... that'd be awesome :)

Ps. I haven't told yet but thanks to create this cool & promising parser!

**Comment by maxbrunsfeld** [comment]
Closing this out because property sheets have been removed, in favor of queries (#444).
